File: MicroCli/src/main/resources/micronaut/security/java/mongoReactive/repository/UserRepository.java
package io.hashimati.security.repository;


import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoCollection;
import io.hashimati.security.domains.User;
import jakarta.inject.Singleton;
import org.bson.BsonDocument;
import org.bson.BsonString;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Singleton
public class UserRepository
{
    private static final Logger log = LoggerFactory.getLogger(UserRepository.class);

    private final MongoClient mongoClient;
    private MongoCollection<User> collection;
    public UserRepository(MongoClient mongoClient){

        this.mongoClient = mongoClient;
        log.info("Creating Collection for UserRepository");
        collection = mongoClient.getDatabase("database")
                .getCollection("users", User.class);
    }

    private MongoCollection<User> getCollection()
    {
        return collection;
    }

    private Mono<User> findAsMono(BsonDocument query)
    {
        log.info("Finding query : {}", query);
        return Mono.from(getCollection()
                .find(query));
    }

    private Flux<User> findAsFlux(BsonDocument query)
    {
        log.info("Finding query : {}", query);
        return Flux
                .from(getCollection()
                        .find(query));
    }


    public Mono<User> save(User object)
    {

        log.info("Saving User : {}", object);
        if(object.getId() == null)
        {
            return Mono.just(object)
                    .onErrorReturn(object);
        }
        return Mono.from(getCollection().insertOne(object)
        ).map(success ->object);
    }

    public Flux<User> findAll()
    {
        log.info("Finding All Users ");
        return Flux.from(getCollection().find());
    }

    public Mono<User> findById(final String id){
        log.info("Finding Fruit by Id : {}", id);
        final BsonDocument document = new BsonDocument();
        document.append("_id", new BsonString(id));
        return findAsMono(document);
    }

    public Mono<Boolean> deleteById(String id) {
        log.info("Deleting User by id: {}", id);
        BsonDocument document = new BsonDocument();
        document.append("_id", new BsonString(id));
        return Mono.from(getCollection().deleteOne(document))
                .map(success->Boolean.TRUE)
                .onErrorReturn(Boolean.FALSE);
    }

    public Mono<User> update(User object)  {
        log.info("update user {}", object.getUsername() );
        BsonDocument document = new BsonDocument();
        document.append("_id", new BsonString(object.getId()));
        return  Mono.from(getCollection().findOneAndReplace(document,object ));

    }

    public Mono<User> findByUsername(String username) {
        log.info("Find user by username {}", username);
        BsonDocument document = new BsonDocument();
        document.append("username", new BsonString(username));
        return  findAsMono(document);

    }

    public boolean existsByUsername(String username) {
        log.info("Exist {}", username);
        return findByUsername(username).blockOptional().isPresent();
    }
}

File: Code Examples/JWT/Kotlin/JWTKotlinMongo/src/main/kotlin/io/hashimati/security/repository/UserRepository.kt
package io.hashimati.security.repository

import com.mongodb.client.result.DeleteResult
import com.mongodb.client.result.InsertOneResult
import com.mongodb.reactivestreams.client.MongoClient
import com.mongodb.reactivestreams.client.MongoCollection
import io.hashimati.security.domains.User
import jakarta.inject.Singleton
import org.bson.BsonDocument
import org.bson.BsonString
import org.slf4j.LoggerFactory
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono


@Singleton
class UserRepository(private val mongoClient: MongoClient) {
    private val collection: MongoCollection<User>

    private fun findAsMono(query: BsonDocument): Mono<User> {
        log.info("Finding query : {}", query)
        return Mono.from(
            collection
                .find(query)
        )
    }

    private fun findAsFlux(query: BsonDocument): Flux<User> {
        log.info("Finding query : {}", query)
        return Flux
            .from(
                collection
                    .find(query)
            )
    }

    fun save(`object`: User): Mono<User> {
        log.info("Saving User : {}", `object`)
        return if (`object`.id == null) {
            Mono.just(`object`)
                .onErrorReturn(`object`)
        } else Mono.from(
            collection.insertOne(`object`)
        ).map { success: InsertOneResult? -> `object` }
    }

    fun findAll(): Flux<User> {
        log.info("Finding All Users ")
        return Flux.from(collection.find())
    }

    fun findById(id: String?): Mono<User> {
        log.info("Finding Fruit by Id : {}", id)
        val document = BsonDocument()
        document.append("_id", BsonString(id))
        return findAsMono(document)
    }

    fun deleteById(id: String?): Mono<Boolean> {
        log.info("Deleting User by id: {}", id)
        val document = BsonDocument()
        document.append("_id", BsonString(id))
        return Mono.from(collection.deleteOne(document))
            .map { success: DeleteResult? -> java.lang.Boolean.TRUE }
            .onErrorReturn(java.lang.Boolean.FALSE)
    }

    fun update(`object`: User): Mono<User> {
        log.info("update user {}", `object`.username)
        val document = BsonDocument()
        document.append("_id", BsonString(`object`.id))
        return Mono.from(collection.findOneAndReplace(document, `object`))
    }

    fun findByUsername(username: String?): Mono<User> {
        log.info("Find user by username {}", username)
        val document = BsonDocument()
        document.append("username", BsonString(username))
        return findAsMono(document)
    }

    fun existsByUsername(username: String?): Boolean {
        log.info("Exist {}", username)
        return findByUsername(username).blockOptional().isPresent
    }

    companion object {
        private val log = LoggerFactory.getLogger(UserRepository::class.java)
    }

    init {
        log.info("Creating Collection for UserRepository")
        collection = mongoClient.getDatabase("database")
            .getCollection("users", User::class.java)
    }
}